yetanotherco/aligned_layer
# Fix: Batcher User Balance Checks

# Fix: Batcher User Balance Checker

## Description

Replaces the hardcoded constant in https://github.com/yetanotherco/aligned_layer/blob/testnet/batcher/aligned-batcher/src/lib.rs#L607 with the user's `min_fee`. Addresses the first point of #1383 . 

This fix is done by adding `total_fees_in_queue` variable in the UserState struct, and comparing if user has enough balance to pay for the needed amount.

This PR also refactors the infamous `min_fee` variable, to be called `max_fee_limit` / `last_max_fee_limit` 

This PR also adds some metrics of this. dont forget to visualize them in the following tests running
```
make run_metrics
```

### To Test:
#### test 1
- Run a local network and ensure proofs are being validated
- Submit a single proof with no balance in the BatcherPaymentService Contract :
1.) Start a local testnet with an no balance:
2.) Create a wallet with no funds
```sh
cast wallet new <KEYSTORE_PATH>
```
3.)
```bash
echo "Sending SP1 fibonacci task to Batcher..."
cd batcher/aligned/ && cargo run --release -- submit \
		--proving_system SP1 \
		--proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
		--vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
		--repetitions 1 \
		--proof_generator_addr <WALLET_ADDRESS> \
		--keystore_path <KEYSTORE_PATH> \
		--rpc_url http://localhost:8545 \
		--network devnet
```
ensure the proof submission errors within the batcher with `InsufficientBalance`.

#### test 2
- Submit multiple proofs and not have enough to pay.
1.) Fund the wallet by sending eth from a prefunded anvil account:

```bash
cast send <WALLET_ADDRESS --rpc-url http://localhost:8545 --private-key 2a871d0798f97d79848a013d4936a73bf4cc922c825d33c1cf7073dff6d409c6 --value 1ether
```
2.) Fund the batcher with an insufficient balance for multiple proofs.
```bash
aligned deposit-to-batcher \
--rpc_url http://localhost:8545 \                      
--network devnet \ 
--keystore_path <KEYSTORE_PATH> \
--amount 0.0026ether
```
3.) Send one proof, see it succeeds.
```bash
cargo run --release -- submit \
                --proving_system SP1 \
                --proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
                --vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
                --repetitions 1 \
                --proof_generator_addr <WALLET_ADDRESS \
                --keystore_path <KEYSTORE_PATH> \
                --rpc_url http://localhost:8545 \
                --network devnet
```
4.) Send a burst of 10 or more proofs, observe it submission fails due to insufficient balance.
```bash
cargo run --release -- submit \
                --proving_system SP1 \
                --proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
                --vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
                --repetitions 10 \
                --proof_generator_addr <WALLET_ADDRESS \
                --keystore_path <KEYSTORE_PATH> \
                --rpc_url http://localhost:8545 \
                --network devnet
```
![Uploading Screenshot 2024-11-13 at 14.03.31.png…]()

## Type of change

Please delete options that are not relevant.


## Checklist
